File: kangtaku/search_in_rotated_sorted_array.java
class Solution {
    public int search(int[] nums, int target) {
        if(nums.length==0)return -1;
        int n=nums.length,lo=0,hi=n-1;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(nums[mid]==target)return mid;
            if(nums[0]>nums[n-1]&&target<=nums[n-1]&&nums[n-1]<=nums[mid])lo=mid+1;
            else if(nums[0]>nums[n-1]&&target>=nums[0]&&nums[n-1]>=nums[mid])hi=mid-1;
            else if(nums[mid]>target)hi=mid-1;
            else if(nums[mid]<target)lo=mid+1;
        }
        return -1;
    }
}

File: iamminji/leetcode/medium/search_in_rotated_sorted_array.py
# 33. Search in Rotated Sorted Array
# https://leetcode.com/problems/search-in-rotated-sorted-array/description/


class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """

        lo = 0
        hi = len(nums) - 1
        while lo < hi:
            mid = (lo - hi) // 2 + hi
            if nums[mid] == target:
                return mid

            if nums[lo] <= nums[mid]:
                if nums[lo] <= target <= nums[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1
            else:
                if nums[mid] < target <= nums[hi]:
                    lo = mid + 1
                else:
                    hi = mid - 1

        if lo == hi and nums[lo] == target:
            return lo
        return -1
